### PR TITLE
Worker V4 package update

### PR DESCRIPTION
**Pull request by SimonRichardson**
Brings in the worker v4 package, this incudes the changes to the start function
on the manifold config. The new start function now offers a context.Context.
This context.Context is a context for only the duration of the start function,
not the entire life of the worker. This context is used to cancel the start
function if the worker is shutting down.

There might be a possibility in the future to have a Fork function that allows 
the passing of values of a context, that negates the canceling of the
context.

Lastly, I need to fix the dependency test package to update the old naming of
context. That can be done in a separate PR.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Quick sanity check:

```
$ juju bootstrap lxd test --build-agent
$ juju enable-ha
$ juju deploy ubuntu
```

## Links

**Jira card:** JUJU-5286

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -82,7 +82,7 @@ require (
 	github.com/juju/version/v2 v2.0.1
 	github.com/juju/viddy v0.0.0-beta5
 	github.com/juju/webbrowser v1.0.0
-	github.com/juju/worker/v3 v3.5.0
+	github.com/juju/worker/v4 v4.0.0-20240112112653-6268855e643c
```
Comment (wallyworld) [inline go.mod:85]: This should just be a semantic version without the sha

**Review by wallyworld — APPROVED** [review]
Another step forward \o/

File: agent/engine/housing.go [hunk 2]
```diff
@@ -80,15 +82,15 @@ func maybeAdd(original []string, add string) []string {
...
-		if err := context.Get(name, &guest); err != nil {
+		if err := getter.Get(name, &guest); err != nil {
 			return nil, errors.Trace(err)
 		}
 		task := func() (worker.Worker, error) {
-			return inner(context)
+			return inner(ctx, getter)
 		}
-		worker, err := fortress.Occupy(guest, task, context.Abort())
+		worker, err := fortress.Occupy(guest, task, ctx.Done())
```
Comment (hpidcock) [inline agent/engine/housing.go:93]: I'm sure the std `ctx.Done()` here is functionally equivalent to the old`context.Abort()`.

File: cmd/jujud/agent/machine/manifolds.go [hunk 3]
```diff
@@ -665,7 +666,7 @@ func commonManifolds(config ManifoldsConfig) dependency.Manifolds {
 		}),
 
 		charmhubHTTPClientName: dependency.Manifold{
-			Start: func(_ dependency.Context) (worker.Worker, error) {
+			Start: func(_ context.Context, getter dependency.Getter) (worker.Worker, error) {
```
Comment (hpidcock) [inline cmd/jujud/agent/machine/manifolds.go:669]: getter is just as unused here as the context, either both should be unnamed or both named.

File: cmd/jujud/agent/machine/manifolds.go [hunk 4]
```diff
@@ -1125,7 +1126,7 @@ func mergeManifolds(config ManifoldsConfig, manifolds dependency.Manifolds) depe
 
 func clockManifold(clock clock.Clock) dependency.Manifold {
 	return dependency.Manifold{
-		Start: func(_ dependency.Context) (worker.Worker, error) {
+		Start: func(_ context.Context, getter dependency.Getter) (worker.Worker, error) {
```
Comment (hpidcock) [inline cmd/jujud/agent/machine/manifolds.go:1129]: same here

File: cmd/jujud/agent/model/manifolds.go [hunk 5]
```diff
@@ -585,7 +585,7 @@ func CAASManifolds(config ManifoldsConfig) dependency.Manifolds {
 // clockManifold expresses a Clock as a ValueWorker manifold.
 func clockManifold(clock clock.Clock) dependency.Manifold {
 	return dependency.Manifold{
-		Start: func(_ dependency.Context) (worker.Worker, error) {
+		Start: func(_ context.Context, getter dependency.Getter) (worker.Worker, error) {
```
Comment (hpidcock) [inline cmd/jujud/agent/model/manifolds.go:588]: same here, there a more, but please make them all consistent.

File: internal/worker/apiserver/manifold.go [hunk 6]
```diff
@@ -257,7 +258,7 @@ func (config ManifoldConfig) start(context dependency.Context) (worker.Worker, e
 		return nil, errors.Trace(err)
 	}
 
-	w, err := config.NewWorker(stdcontext.TODO(), Config{
+	w, err := config.NewWorker(ctx, Config{
```
Comment (hpidcock) [inline internal/worker/apiserver/manifold.go:261]: Love seeing a TODO go bye bye

**Review by hpidcock — APPROVED** [review]
Nice improvement.

File: go.mod (hunk 1 above)
Comment (SimonRichardson) [reply go.mod:85]: I'll make a release.

File: agent/engine/housing.go (hunk 2 above)
Comment (SimonRichardson) [reply agent/engine/housing.go:93]: In this context it does, I had to change the `worker.Worker` package to ensure it was.

File: cmd/jujud/agent/machine/manifolds.go (hunk 3 above)
Comment (SimonRichardson) [reply cmd/jujud/agent/machine/manifolds.go:669]: grep + sed didn't know what to do here in the original changes, I'll fix as many as I can.

**Pull request merged**